tests/test_unit/test_kinematics.py

```diff
@@ -1,112 +1,176 @@
-        ds = request.getfixturevalue(ds)
-        with expected_exception:
-            result = kinematics.compute_velocity(ds.position)
-            expected = expected_dataarray("velocity")
-            if ds.position.isnull().any():
-                expected.loc[
-                    {"individuals": "ind1", "time": [2, 4, 6, 7, 8, 9]}
-                ] = np.nan
-            xr.testing.assert_allclose(result, expected)
-
-    @pytest.mark.parametrize("ds, expected_exception", kinematic_test_params)
-    def test_acceleration(
-        self, ds, expected_exception, expected_dataarray, request
-    ):
-        """Test acceleration computation."""
-        ds = request.getfixturevalue(ds)
-        with expected_exception:
-            result = kinematics.compute_acceleration(ds.position)
-            expected = expected_dataarray("acceleration")
-            if ds.position.isnull().any():
-                expected.loc[
-                    {"individuals": "ind1", "time": [1, 3, 5, 6, 7, 8, 9]}
-                ] = np.nan
-            xr.testing.assert_allclose(result, expected)
-
-    @pytest.mark.parametrize("order", [0, -1, 1.0, "1"])
-    def test_approximate_derivative_with_invalid_order(self, order):
-        """Test that an error is raised when the order is non-positive."""
-        data = np.arange(10)
-        expected_exception = (
-            ValueError if isinstance(order, int) else TypeError
+@pytest.mark.parametrize(
+    "valid_dataset_with_nan",
+    [
+        "valid_poses_dataset_with_nan",
+        "valid_bboxes_dataset_with_nan",
+    ],
+)
+@pytest.mark.parametrize(
+    "kinematic_variable, expected_nans_per_individual",
+    [
+        ("displacement", {0: 5, 1: 0}),
+        ("velocity", {0: 6, 1: 0}),
+        ("acceleration", {0: 7, 1: 0}),
+    ],
+)
+def test_kinematics_with_dataset_with_nans(
+    valid_dataset_with_nan,
+    kinematic_variable,
+    expected_nans_per_individual,
+    helpers,
+    request,
+):
+    """Test kinematics computation for a dataset with nans.
+
+    We test that the kinematics can be computed and that the number
+    of nan values in the kinematic array is as expected.
+
+    """
+    # compute kinematic array
+    valid_dataset = request.getfixturevalue(valid_dataset_with_nan)
+    position = valid_dataset.position
+    kinematic_array = getattr(kinematics, f"compute_{kinematic_variable}")(
+        position
+    )
+
+    # compute n nans in kinematic array per individual
+    n_nans_kinematics_per_indiv = {
+        i: helpers.count_nans(kinematic_array.isel(individuals=i))
+        for i in range(valid_dataset.dims["individuals"])
+    }
+
+    # check number of nans per indiv is as expected in kinematic array
+    for i in range(valid_dataset.dims["individuals"]):
+        assert n_nans_kinematics_per_indiv[i] == (
+            expected_nans_per_individual[i]
+            * valid_dataset.dims["space"]
+            * valid_dataset.dims.get("keypoints", 1)
         )
```

Comment:
```suggestion
    # compute n nans in kinematic array per individual
    n_nans_kinematics_per_indiv = [
        helpers.count_nans(kinematic_array.isel(individuals=i))
        for i in range(valid_dataset.dims["individuals"])
    ]
    # expected nans per individual adjusted for space and keypoints dimensions
    expected_nans_adjusted = [
        n
        * valid_dataset.dims["space"]
        * valid_dataset.dims.get("keypoints", 1)
        for n in expected_nans_per_individual
    ]
    # check number of nans per individual is as expected in kinematic array
    np.testing.assert_array_equal(
        n_nans_kinematics_per_indiv, expected_nans_adjusted
    )
```
And we can assert the array as a whole

Reply:
looks so clear, thanks! ✨ 🤩